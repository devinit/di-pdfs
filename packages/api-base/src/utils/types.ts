import {IDatabase} from 'pg-promise';
import {IExtensions} from '../db';

export interface IGetIndicatorArgs {
    id?: string;
    query: string;
    country?: string;
    table?: string;
    conceptType: string; // folder with concept file
    db: IDatabase<IExtensions> & IExtensions;
}
export interface ISpotlightGetIndicatorArgs {
    id: string;
    query: string;
    country: string;
    l1?: string;
    schema: string;
    table?: string;
    conceptType: string; // folder with concept file
    db: IDatabase<IExtensions> & IExtensions;
}
export interface IRAWPopulationGroup {
    di_id: string;
    value_rural: string;
    value_urban: string;
    year: string;
}
export interface IRAWPopulationAgeBand {
    di_id: string;
    value_0_14: string;
    value_15_64: string;
    value_65_and_above: string;
    year: string;
}
export interface IRAW {
    di_id: string;
    value_ncu?: string;
    district_id: string;
    value: string;
    year: string;
}
export interface IProcessedSimple {
    id: string;
    value: number;
    year: number;
    uid: string;
    budget_type?: string;
}
export interface IRAWMulti {
    di_id: string;
    value: string;
    value_1: string;
    year: string;
}
export interface IRAWQuintile {
    value_bottom_20pc: string;
    value_2nd_quintile: string;
    value_3rd_quintile: string;
    value_4th_quintile: string;
    value_5th_quintile: string;
}
export interface IRAWFlow {
    di_id: string;
    year: string;
    flow_category_order: number;
    direction: string;
    flow_type: string;
    flow_name: string;
    value: string;
}
export interface IRAWDomestic {
    di_id: string;
    year: string;
    budget_type: string;
    l1: string;
    l2: string;
    l3: string;
    l4: string;
}
export interface IToolTipArgs {
    query?: string;
    id?: string;
    conceptType: string;
}
export interface Isummable {
    value: number | string | null;
}

export interface IhasDiId {
    di_id: string;
}

export interface IhasId {
    id: string | null;
}
export interface IGetIndicatorValueArgs {
    id: string;
    sqlList: string[];
    precision?: number;
    db: IDatabase<IExtensions> & IExtensions;
    format: boolean;
}
