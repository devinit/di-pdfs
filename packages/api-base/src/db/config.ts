import {TConfig} from 'pg-promise';

// get the connection details from Allan or Donata

export const dwConfig: TConfig = {
    user: process.env.DB_USER,
    database: process.env.DB_DATABASE,
    host: process.env.DB_HOST,
    port: Number(process.env.DB_PORT),
    password: process.env.DB_PASSWORD
};
