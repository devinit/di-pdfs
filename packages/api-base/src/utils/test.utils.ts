// a patch for pretty format output that will exclude uids from storyshots
// this is because they are not constants and change on every new storyshots
// the issue with this is that we cant compare new and old storyshots
import * as prettyFormat from 'pretty-format';

export type ListWithUid = Array<{uid: string} & any>;

export interface IHasUid {
    [field: string]: {
        data: ListWithUid;
        toolTip: any;
    };
}

export const replaceUidInList = (data: ListWithUid): ListWithUid => {
    return data.map(obj => {
        if (obj && obj.uid) obj.uid = 'unique_id_stub';
        return obj;
    });
};

export const uidPatchForTabData = (data: IHasUid) => {
    return Object.keys(data).reduce((acc, key: string) => {
        const itemsWithUid: ListWithUid = data[key].data;
        const newItems =  replaceUidInList(itemsWithUid);
        return {...acc, [key]: { ...data[key], data: newItems}};
    }, {});
};

export const prettyLists =
    (data: ListWithUid) => prettyFormat(replaceUidInList(data));
