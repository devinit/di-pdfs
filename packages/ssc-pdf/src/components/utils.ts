import dvptColorMap from '@devinit/ssc-api/lib/modules/shared/config';

type ILegendData = Array<{color: string; value: string | number}>;
interface ILegendOptions {
    legendData: object;
    data?: any[];
    keyName?: string;
}

export const getLegendData = ({legendData, data, keyName}: ILegendOptions): ILegendData => {
    // keys with no data
    const keysToRemove = data && keyName ? data.reduce((all, obj) => {
        if (obj.value === 0) return [...all, obj[keyName]];
        return all;
    }, []) : [];
    return Object.keys(legendData)
        .filter(key => !keysToRemove.includes(key))
        .map(value => ({color: legendData[value], value}));
};

export const dvptCoLegendData = (country: string): Array<{color: string; value: string}> => {
    return Object.keys(dvptColorMap[country]).map(value => {
        const color = dvptColorMap[country][value];
        return {color, value};
    });
};
