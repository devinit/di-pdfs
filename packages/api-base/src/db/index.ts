import {IMain, IDatabase, IOptions} from 'pg-promise';
import {dwConfig} from './config';
import * as pgPromise from 'pg-promise';
import * as LRU from 'lru-cache';

export interface IExtensions {
    manyCacheable: (query: string, values?: object) => Promise<any[]>;
}

const lruOpts: LRU.Options<any> = {
    max: 300,
    maxAge: 1000 * 60 * 60 * 60
};

export const dbCache: LRU.Cache<any, any> = LRU(lruOpts);

// pg-promise initialization options:
const options: IOptions<IExtensions> = {
    // Extending the database protocol with our custom modules
    // API: http://vitaly-t.github.io/pg-promise/global.html#event:extend
    extend: (obj: IExtensions & IDatabase<IExtensions>) => {
        obj.manyCacheable = (query: string, values?: any) => {
            const getQuery = values ? pgPromise.as.format(query, values) : query;
            if (dbCache.has(getQuery)) {
                return Promise.resolve(dbCache.get(getQuery));
            }
            return obj.any(getQuery);
        };
    },
    // caching
    receive: (data, _result, event) => {
        // cache recieved
        if (!dbCache.has(event.query)) dbCache.set(event.query, data);
    }
};

const pgp: IMain = pgPromise(options);

// Create the database instance with extensions:
const db = pgp<IExtensions>(dwConfig);

process.on('exit', (code) => {
  // kill db
  pgp.end();
  console.log(`About to exit with code: ${code}, closed DB connection`);
  process.kill(process.pid);
});

process.on('SIGINT', () => {
    // kill db
    pgp.end();
    console.log('Ctrl-C... forced termination closed DB connection');
    process.kill(process.pid);
});

export type IDB = IDatabase<IExtensions> & IExtensions; // necessary repition to make typescript happy
// If you ever need access to the library's root (pgp object), you can do it via db.$config.pgp
// See: http://vitaly-t.github.io/pg-promise/Database.html#.$config
export default db;
