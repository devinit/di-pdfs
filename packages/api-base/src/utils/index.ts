import * as R from 'ramda';
import {IEntity, ICurrency, getCurrency, getEntityBySlugAsync} from '../cms/modules/global';
import * as shortid from 'shortid';
import {IhasDiId, Isummable} from './types';
import {IDatabase} from 'pg-promise';
import {IExtensions} from '../db';

export type IDB = IDatabase<IExtensions> & IExtensions;

export const RECIPIENT = 'recipient';
export const DONOR = 'donor';
export const MULTILATERAL = 'multilateral';
export const  CROSSOVER = 'crossover';
// TODO: should get from a refrence file
export const  budgetTypesRefs = {
    actual: 'Actual',
    estimated: 'Estimated',
    proposed: 'Proposed',
    approved: 'Approved',
    budget: 'Budget',
    proj: 'Projected'
};

// makes numerical like values numbers in an object
export const toNumericFields: (obj: any) => any = (obj) => {
    return R.keys(obj).reduce((newObj: any, key: string) => {
        const isKeyNumerical = Number(obj[key]) ? true : false;
        if (isKeyNumerical) {
            return {...newObj, [key]:  Number(obj[key])};
        }
        return {...newObj, [key]: obj[key]};
    }, {});
};

export const toId: (obj: IhasDiId ) => any = (obj) => {
    if (!obj.di_id) return obj;
    const id = obj.di_id;
    const newObj = R.omit(['di_id'], obj);
    return {...newObj, id };
};

export const getTotal = (data: Isummable[]): number =>
    R.reduce<Isummable, number>((sum: number, obj: Isummable): number => {
        if (obj.value) sum += Number(obj.value);
        return sum;
    }, 0, data);

export const valuesIntoPercents = <T extends {value: number}>(data: T[], precision?: number): T[] => {
    const sum = getTotal(data);
    return data.map((obj: T) => {
        const newValue  = (obj.value / sum) * 100;
        const value = precision ? Number(formatNumbers(newValue, precision)) : Math.round(newValue);
        return Object.assign(obj, {value});
    });
};

export const groupedValuesIntoPercents =
    <T extends {value: number}>(data: T[], groupName: string, precision?: number): T[] => {
    const groups = R.groupBy(R.prop(groupName), data);
    const groupAsPers: any[][] = R.keys(groups).map(key =>
        valuesIntoPercents(groups[key], precision));
    return R.flatten(groupAsPers);
};

export const getCurrencyCode = async (id: string): Promise<string>  => {
    try {
        const currencyList: ICurrency[] = await getCurrency();
        const entity: IEntity | undefined = await getEntityBySlugAsync(id);
        if (!entity) throw new Error(`entity was not found for slug: ${id}`);
        const currency: ICurrency | undefined = R.find(R.propEq('id', entity.id), currencyList) as ICurrency;
        return currency ? currency.code : 'NCU';
    } catch (error) {
        throw error;
    }
};

export const getTableNameFromSql = (sqlStr: string): string | Error => {
    const matches = sqlStr.match(/FROM(.*)WHERE/);
    if (matches && matches.length) {
        return matches[0].split(/\s/)[1];
    }
    return new Error(`couldnt get table name from sql string ${sqlStr}`);
};

export const getIndicatorDataSimple = async<T extends {}> ({db, query}: {db: IDB, query: string}): Promise<T[]> => {
    const raw = await db.manyCacheable(query);
    return raw
        .map(toNumericFields)
        .map(obj => ({...obj, uid: shortid.generate()})) as T[];
};

export const isDonor = async (slug: string): Promise<boolean>  => {
    const obj: IEntity | undefined = await getEntityBySlugAsync(slug);
    if (!obj) throw new Error('Error in isDonor function, entity is undefined');
    if (obj.donor_recipient_type === DONOR) return true;
    return false;
};

export const normalizeKeyName = (columnName: string): string => {
    const str = columnName.includes('value_') ? columnName.split(/value\_/)[1] : columnName;
    return str.replace(/\_/g, '-');
};

export const normalizeKeyNames = (obj: {}) => {
    return R.keys(obj).reduce((acc, key) => {
        const newKeyName = key.includes('_') ? normalizeKeyName(key) : key;
        if (newKeyName) {
            const newObj = R.omit([key], obj);
            return {...newObj, [newKeyName]: obj[newKeyName]};
        }
        return {...acc, [key]: obj[key]}; // return to default
    }, {});
};

export const getCurrentYear = (): number => {
    const date = new Date();
    return date.getFullYear();
};

export const capitalize = (val: string) =>
    `${val[0].toUpperCase()}${R.drop(1, val)}`;

export const addSuffix = (val) => {
    // Borrowed from old datahub codebase
    const lastDigit = (val % 10);
    const lastTwoDigits = (val % 100);
    const suffixObj = ['th', 'st', 'nd', 'rd', 'th', 'th', 'th', 'th', 'th', 'th'];
    const exceptions = [11, 12, 13];
    if (exceptions.indexOf(lastTwoDigits) === -1) {
        return val + suffixObj[lastDigit];
    }
    return val + 'th';
};

export const getMaxAndMin = (data: Array<{year?: number | null}>): number[] => {
    const years = data
        .map(obj => {
            if (obj && obj.year) return Number(obj.year);
            return null;
        })
        .filter(year => year !== null);
    if (!years) return [0, 0];
    const max: number = Math.max.apply(null, years);
    const min: number = Math.min.apply(null, years);
    return [max, min];
};

const removeTrailingZero = (value: string): string => {
    const val = Number(value);
    return  Math.round(val) === val ? val.toString() : value;
};

// (10 ** length) == Math.pow(10, length);
export const roundNum = (num, length): string =>
    (Math.round(num * (10 ** length)) / (10 ** length)).toFixed(length);

export const formatNumbers =
    (value: number | string | undefined | null,
     precision: number = 1,
     shouldrRemoveTrailingZero: boolean = false): string => {
    if (value === undefined || value === null) return 'No data';
    const val = Number(value);
    const absValue = Math.abs(val);
    if (absValue < 1e3) {
        const fixed = roundNum(val, precision);
        return shouldrRemoveTrailingZero ? `${removeTrailingZero(fixed)}` : fixed;
    } else if (absValue >= 1e3 && absValue < 1e6) {
        const newValue = val / 1e3;
        const fixed = roundNum(newValue, precision);
        return shouldrRemoveTrailingZero ? `${removeTrailingZero(fixed)}k` : `${fixed}k`;
    } else if (absValue >= 1e6 && absValue < 1e9) {
        const newValue = val / 1e6;
        const fixed = roundNum(newValue, precision);
        return shouldrRemoveTrailingZero ? `${removeTrailingZero(fixed)}m` : `${fixed}m`;
    } else {
        const newValue = val / 1e9;
        const fixed = roundNum(newValue, precision);
        return shouldrRemoveTrailingZero ? `${removeTrailingZero(fixed)}bn` : `${fixed}bn`;
    }
};
