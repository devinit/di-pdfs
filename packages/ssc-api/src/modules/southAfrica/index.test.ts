import {prettyLists} from '@di-pdfs/api-base/lib/utils/test.utils';
import {SouthAfrica} from '.';
import db from '@di-pdfs/api-base/lib/db';

describe('SouthAfrica', () => {
    const southAfrica = new SouthAfrica(db);
    it('getting dvptCooperationTrend', async () => {
        const data = await southAfrica.dvptCooperationTrend();
        expect(prettyLists(data)).toMatchSnapshot();
     }, 100000);
    it('getting aricfExpByRegion', async () => {
        const data = await southAfrica.aricfExpByRegion();
        expect(prettyLists(data)).toMatchSnapshot();
     }, 100000);
    it('getting govmtdepartment', async () => {
        const data = await southAfrica.govmtdepartment();
        expect(prettyLists(data)).toMatchSnapshot();
     }, 100000);
    it('getting aricfExpBySector', async () => {
        const data = await southAfrica.aricfExpBySector();
        expect(prettyLists(data)).toMatchSnapshot();
     }, 100000);
    afterAll(() => {
       db.$config.pgp.end();
    });
});
