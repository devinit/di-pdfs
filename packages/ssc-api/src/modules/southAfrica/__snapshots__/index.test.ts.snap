// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SouthAfrica getting aricfExpByRegion 1`] = `
"Array [
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"region\\": \\"Africa\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 11.4,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#EA7600\\",
    \\"region\\": \\"Asia\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 0.7,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"region\\": \\"Africa\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 6.7,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#EA7600\\",
    \\"region\\": \\"Asia\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 0.2,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#93328E\\",
    \\"region\\": \\"Latin America and the Caribbean\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 6.5,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"region\\": \\"Africa\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 4.3,
    \\"year\\": 2016,
  },
  Object {
    \\"color\\": \\"#93328E\\",
    \\"region\\": \\"Latin America and the Caribbean\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 2.4,
    \\"year\\": 2016,
  },
]"
`;

exports[`SouthAfrica getting aricfExpBySector 1`] = `
"Array [
  Object {
    \\"color\\": \\"#E03997\\",
    \\"sector\\": \\"ARICF expenditures (total)\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 50.1,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"sector\\": \\"Coopertion between South Africa and other states\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 15.3,
  },
  Object {
    \\"color\\": \\"#FF8EDF\\",
    \\"sector\\": \\"Prevention and resolution of conflicts\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 2.5,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"sector\\": \\"Socio-economic  development and integration\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 2.8,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"sector\\": \\"Promotion of democracy and good governance\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 4.7,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"sector\\": \\"Humanitarian assistance\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 24.3,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"sector\\": \\"Human resources development\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 0.3,
  },
]"
`;

exports[`SouthAfrica getting dvptCooperationTrend 1`] = `
"Array [
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 293,
    \\"year\\": 2015,
  },
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 174,
    \\"year\\": 2014,
  },
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 182,
    \\"year\\": 2013,
  },
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 176,
    \\"year\\": 2012,
  },
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 155,
    \\"year\\": 2011,
  },
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 151,
    \\"year\\": 2010,
  },
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 193,
    \\"year\\": 2009,
  },
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 150,
    \\"year\\": 2008,
  },
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 112,
    \\"year\\": 2007,
  },
  Object {
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 99,
    \\"year\\": 2006,
  },
]"
`;

exports[`SouthAfrica getting govmtdepartment 1`] = `
"Array [
  Object {
    \\"color\\": \\"#DC73FF\\",
    \\"id\\": \\"International Relations and Cooperation\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 69.18604651162791,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"National Treasury\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 20.994832041343674,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 9.819121447028424,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#DC73FF\\",
    \\"id\\": \\"International Relations and Cooperation\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 65.70130607609312,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"National Treasury\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 25.04258943781942,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 9.25610448608745,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#DC73FF\\",
    \\"id\\": \\"International Relations and Cooperation\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 66.66666666666666,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"National Treasury\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 24.546952224052717,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 8.786381109280613,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#DC73FF\\",
    \\"id\\": \\"International Relations and Cooperation\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 63.96085204375359,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"National Treasury\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 27.461139896373055,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 8.578008059873344,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#DC73FF\\",
    \\"id\\": \\"International Relations and Cooperation\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 35.195911413969334,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"National Treasury\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 59.557069846678026,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 5.247018739352641,
    \\"year\\": 2015,
  },
]"
`;
